### ISSUE TITLE
Should warn about nullability of assigning `default` to a struct with non-nullable fields

### ISSUE THREAD

**Issue by jcouv** [dotnet/roslyn#32179]
```C#
struct S
{
    string field;
    void M(S parameter)
    { 
        S local = default; // we just set `local.field` to null (we should warn)
        M(default); // we should also warn
    }
}
```

We probably need some discussion on private fields. We could:
1. not warn (opens some safety issues)
2. warn the consumer of the struct
3. warn the producer of the struct (private fields or maybe even all fields should be nullable)

Also @sharwell mentioned some related scenarios:
```C#
class C { private S _field; }
class D
{
   void M()
   {
       var c = new C(); // equivalent to `new C { _field = default }`
   }
}
```

Note: this is one more reason why reference assemblies should either include all fields, or at least keep a "representative" set of fields.

FYI @cston

**Comment by huoyaoyuan** [comment]
```csharp
struct S { public object _field; }
public T M() => default;
var s = M<S>();
```
Which one should warn?

**Comment by jcouv** [comment]
@huoyaoyuan That's a good scenario. I think it pushes us towards option (3), namely warning on the declaration of the struct.

**Comment by jcouv** [comment]
Some notes from discussion with Mads:
The current behavior seems to have been an explicit decision. It's a known hole, similar to array construction (`var x = new string[10];`).
That said we could revisit it.

One concern with warning producers is that it will be annoying (much existing code would warn).
One concern with warning consumers is that there is nothing they can do about warnings about private fields.

**Comment by jcouv** [comment]
From discussion in LDM 2019-03-04, we're keeping this loophole.
